File: examples/subplots/basic.py
#!/usr/bin/env python3
from __future__ import annotations

from pathlib import Path
import argparse
import math

import matplotlib.pyplot as plt
import numpy as np


def save(fig, path):
    fig.savefig(path, dpi=fig.dpi, facecolor=fig.get_facecolor())
    plt.close(fig)
    print(f"saved {path}")


def main():
    parser = argparse.ArgumentParser(description="Matplotlib counterpart for subplots/basic.go")
    parser.add_argument("--out", default="subplots_basic_python.png")
    args = parser.parse_args()
    # Four panels share limits; each varies frequency by column and damping by row.
    fig, axs = plt.subplots(2, 2, figsize=(12, 8), dpi=100, sharex=True, sharey=True)
    fig.subplots_adjust(left=0.08, right=0.96, bottom=0.08, top=0.92, wspace=0.08, hspace=0.08)
    x = np.linspace(0, 10, 128)
    for row in range(2):
        for col in range(2):
            ax = axs[row, col]
            v = np.linspace(0, 1, 128)
            y = np.sin((col + 1) * x) * np.exp(-0.05 * v * (row + 1))
            ax.plot(x, y)
            ax.set_title(f"Panel {row + 1}-{col + 1}")
            ax.set_xlabel("x"); ax.set_ylabel("y"); ax.grid(True)
    axs[0, 0].set_xlim(0, 10); axs[0, 0].set_ylim(-1.2, 1.2)
    save(fig, args.out)


if __name__ == "__main__":
    main()

File: examples/subplots/basic.go
package main

import (
	"fmt"
	"math"

	"matplotlib-go/backends"
	_ "matplotlib-go/backends/all"
	"matplotlib-go/core"
	"matplotlib-go/render"
)

func main() {
	fig := core.NewFigure(1200, 800)

	// Build the same 2x2 shared-axis layout as pyplot.subplots(...,
	// sharex=True, sharey=True) in the Python counterpart.
	grid := fig.Subplots(
		2,
		2,
		core.WithSubplotPadding(0.08, 0.96, 0.08, 0.92),
		core.WithSubplotSpacing(0.08, 0.08),
		core.WithSubplotShareX(),
		core.WithSubplotShareY(),
	)

	for row, rowAxes := range grid {
		for col, ax := range rowAxes {
			ax.SetTitle(fmt.Sprintf("Panel %d-%d", row+1, col+1))
			ax.AddXGrid()
			ax.AddYGrid()

			// Each panel shares x samples but varies frequency by column and
			// damping by row, matching the Python loop structure.
			const n = 128
			x := make([]float64, n)
			y := make([]float64, n)
			for i := 0; i < n; i++ {
				v := float64(i) / float64(n-1)
				xv := 10 * v
				x[i] = xv
				y[i] = math.Sin(float64(col+1)*xv) * math.Exp(-0.05*v*float64(row+1))
			}

			ax.Plot(x, y)
			ax.SetXLabel("x")
			ax.SetYLabel("y")
		}
	}

	// Shared axes are controlled by a single axis; set limits on one axes.
	grid[0][0].SetXLim(0, 10)
	grid[0][0].SetYLim(-1.2, 1.2)

	r, _, createErr := backends.NewRendererFromEnv(backends.Config{
		Width:      1200,
		Height:     800,
		Background: render.Color{R: 1, G: 1, B: 1, A: 1},
		DPI:        96,
	}, backends.TextCapabilities)
	if createErr != nil {
		fmt.Printf("error creating renderer: %v\n", createErr)
		return
	}

	if err := core.SavePNG(fig, r, "subplots_basic.png"); err != nil {
		fmt.Printf("error saving PNG: %v\n", err)
		return
	}

	fmt.Println("saved subplots_basic.png")
}
